### PR TITLE
Refactor actions, separate entities.actions from entities.datasources

### LINKED ISSUE #2164: [Task] Remove datasource data from entities.action redux state
We currently overwrite responses from `/actions` and `/datasources` in our redux state.
This can cause some keys like `datasource.datasourceConfiguration` to go missing in certain cases.
We want to keep our redux state almost* in sync with the server's representation of data.

We want to lookup datasource information from `entities.datasource` on the fly. To make life easier, we'll use a selector.

* - We might change lists to maps for easier lookup.

### PR DESCRIPTION
**Pull request by ofpiyush**
## Description
Makes `entities.datasource` separate from `entities.actions`.

Fixes #2164 

## Type of change
- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Existing tests from #2087 should not break as I undo the hotfixes from #2087 and #2134

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by hetunandu** [comment]
@all-contributors add @ofpiyush for code

**Comment by ofpiyush** [comment]
Hold up on review, I'm moving the eslint changes out

File: app/client/src/entities/Action/index.ts [hunk 1]
```diff
@@ -49,47 +50,67 @@ export interface ApiActionConfig extends ActionConfig {
 }
 
 export interface QueryActionConfig extends ActionConfig {
-  body: string;
+  body?: string;
 }
 
-export interface Action {
+export const isActionDatasource = (val: any): val is ActionDatasource => {
```
Comment (hetunandu) [inline app/client/src/entities/Action/index.ts:56]: isStoredDatasource

File: app/client/src/entities/Action/index.ts [hunk 2]
```diff
@@ -49,47 +50,67 @@ export interface ApiActionConfig extends ActionConfig {
...
+  actionConfiguration: ApiActionConfig;
+}
+
+export interface EmbeddedApiAction extends BaseApiAction {
+  datasource: EmbeddedRestDatasource;
 }
 
-export interface RapidApiAction extends Action {
-  actionConfiguration: Partial<ApiActionConfig>;
+export interface DatasourceApiAction extends BaseApiAction {
```
Comment (hetunandu) [inline app/client/src/entities/Action/index.ts:91]:
```suggestion
export interface StoredDatasourceApiAction extends BaseApiAction {
```

File: app/client/src/entities/Action/index.ts [hunk 3]
```diff
@@ -49,47 +50,67 @@ export interface ApiActionConfig extends ActionConfig {
...
 }
 
-export interface QueryAction extends Action {
-  actionConfiguration: Partial<QueryActionConfig>;
-  eventData?: any;
+export type Action = ApiAction | QueryAction;
+
+export interface ActionWithDatasource {
+  action: Action;
+  datasource: EmbeddedRestDatasource | Datasource;
```
Comment (hetunandu) [inline app/client/src/entities/Action/index.ts:115]: Remove thos

File: app/client/src/api/ActionAPI.tsx [hunk 4]
```diff
@@ -109,42 +110,49 @@ class ActionAPI extends API {
...
   }
 
   static createAPI(
     apiConfig: Partial<Action>,
   ): AxiosPromise<ActionCreateUpdateResponse> {
-    return API.post(ActionAPI.url, apiConfig);
+    const action = Object.assign({}, apiConfig);
+    // In case we were sent an action with an id, delete it before sending to the server
+    // Example: createAPI
+    delete action.id;
```
Comment (hetunandu) [inline app/client/src/api/ActionAPI.tsx:123]: Not needed. Still belongs in copy/move saga

File: app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx [hunk 5]
```diff
@@ -12,9 +12,12 @@ import CodeEditor, {
 import { API_EDITOR_FORM_NAME } from "constants/forms";
 import { AppState } from "reducers";
 import { connect } from "react-redux";
-import { Datasource } from "api/DatasourcesApi";
+import { Datasource } from "entities/Datasource";
```
Comment (hetunandu) [inline app/client/src/components/editorComponents/form/fields/EmbeddedDatasourcePathField.tsx:15]: Bad import

File: app/client/src/entities/DataTree/dataTreeFactory.ts [hunk 6]
```diff
@@ -105,7 +105,7 @@ export class DataTreeFactory {
           }),
         );
       }
-      dataTree[action.config.name] = {
+      const dataTreeAction: DataTreeAction = {
```
Comment (hetunandu) [inline app/client/src/entities/DataTree/dataTreeFactory.ts:108]: Check once again

File: app/client/src/pages/Editor/DataSourceEditor/Connected.tsx [hunk 7]
```diff
@@ -96,18 +101,20 @@ const Connected = () => {
...
-    if (isInOnboarding && showingTooltip === OnboardingStep.EXAMPLE_DATABASE) {
-      payload = {
-        ...payload,
-        name: "ExampleQuery",
-        actionConfiguration: {
-          body: "select * from public.users limit 10",
-        },
-      };
-    }
+    } as Partial<QueryAction>; // TODO: refactor later
```
Comment (hetunandu) [inline app/client/src/pages/Editor/DataSourceEditor/Connected.tsx:104]: Expand on the comments here for TODO

File: app/client/src/pages/Editor/Explorer/hooks.ts [hunk 8]
```diff
@@ -46,7 +47,11 @@ export const useFilteredDatasources = (searchKeyword?: string) => {
   const datasources = useMemo(() => {
     const datasourcesPageMap: Record<string, Datasource[]> = {};
     for (const [key, value] of Object.entries(actions)) {
-      const datasourceIds = value.map((action) => action.config.datasource?.id);
+      const datasourceIds = value.map(
+        (action) =>
+          isActionDatasource(action.config.datasource) &&
+          action.config.datasource.id,
+      );
```
Comment (hetunandu) [inline app/client/src/pages/Editor/Explorer/hooks.ts:54]: Filter first and then map

**Review by hetunandu — DISMISSED** [review]

**Review by hetunandu — APPROVED** [review]

**Pull request merged**